MapboxNavigation/NavigationMapView.swift

```diff
@@ -427,7 +439,60 @@ open class NavigationMapView: MGLMapView {
             }
         }
     }
+    /**
+     Fired when the user taps a waypoint. This recognizer fires a delegate method
+     that prompts the user to remove the waypoint from the route.
+     */
+    func didTapWaypoint(tap: UITapGestureRecognizer) {
+        guard let routes: [Route] = routes else { return }
+        let tapPoint = tap.location(in: self)
+        let tapCoordinate = self.convert(tapPoint, toCoordinateFrom: tap.view)
+        guard let candidates = waypoints(on: routes, closeTo: tapCoordinate) else { return }
+
+        //take the best one, if existant, and prompt to delete
+        guard let best = candidates.1.first, let multipoint = candidates.0.first else { return }
+        requestNew(route: multipoint, without: best)
+    }
+    
+    private func waypoints(on routes: [Route], closeTo coordinate: CLLocationCoordinate2D) -> ([Route], [Waypoint])? {
+        let waypointThreshold: CLLocationDistance = 100
+        let multipointRoutes = routes.filter { $0.routeOptions.waypoints.count >= 3}
+        guard multipointRoutes.count > 0 else { return nil }
+        let waypoints = multipointRoutes.flatMap({$0.routeOptions.waypoints})
+        
+        let closest = waypoints.sorted { (left, right) -> Bool in
+            let leftDistance = left.coordinate.distance(to: coordinate)
+            let rightDistance = right.coordinate.distance(to: coordinate)
+            return leftDistance < rightDistance
+        }
+        let candidates = closest.filter({ $0.coordinate.distance(to: coordinate) < waypointThreshold})
+        return (multipointRoutes, candidates)
+    }
     
+    open override func gestureRecognizerShouldBegin(_ gestureRecognizer: UIGestureRecognizer) -> Bool {
+        if gestureRecognizer == waypointGestureRecognizer {
+            let tapPoint = gestureRecognizer.location(in: self)
+            let tapCoordinate = self.convert(tapPoint, toCoordinateFrom: gestureRecognizer.view)
+            guard let candidates = waypoints(on: routes!, closeTo: tapCoordinate) else { return false }
+            return candidates.1.count > 0
+        }
+        return super.gestureRecognizerShouldBegin(gestureRecognizer)
+    }
+    
+    private func requestNew(route: Route, without waypoint: Waypoint) {
+        let options = route.routeOptions
+        guard let index = options.waypoints.index(of: waypoint) else { return }
+        options.waypoints.remove(at: index)
+        
+        removeRoutes()
+        removeWaypoints()
```

Comment:
No need to remove these; if you call showRoute/waypoint, we update routes/waypoints if they are already on the map.